File: tests/comparison/cpp/string_ops.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <algorithm>

/**
 * String operations benchmark for C++
 * Usage: ./string_ops <iterations>
 */

int string_concatenation(int n) {
    // String concatenation performance - standardized across all languages
    std::string result = "";
    for (int i = 0; i < n; ++i) {
        // Simple character concatenation (matching Caesar implementation)
        result += "x";
    }
    return result.length();
}

int string_processing(int n) {
    // String processing operations - standardized across all languages
    std::string text = "Hello World! This is a test string for benchmarking purposes.";
    int count = 0;
    
    for (int i = 0; i < n; ++i) {
        // Simple string length operation (matching Caesar implementation)
        int length = text.length();
        count += length;
    }
    
    return count;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <iterations>" << std::endl;
        return 1;
    }
    
    int n = std::atoi(argv[1]);
    if (n < 0) {
        std::cerr << "Error: iterations must be non-negative" << std::endl;
        return 1;
    }
    
    // Choose test based on scale
    int result;
    if (n <= 10000) {
        result = string_concatenation(n);
    } else {
        result = string_processing(n / 100);
    }
    
    // Don't print result to avoid affecting timing
    
    return 0;
}

File: tests/comparison/python/string_ops.py
#!/usr/bin/env python3
"""
String operations benchmark for Python
Usage: python string_ops.py <iterations>
"""

import sys

def string_concatenation(n):
    """String concatenation performance - standardized across all languages"""
    result = ""
    for i in range(n):
        # Simple character concatenation (matching Caesar implementation)
        result += "x"
    return len(result)

def string_processing(n):
    """String processing operations - standardized across all languages"""
    text = "Hello World! This is a test string for benchmarking purposes."
    count = 0
    
    for _ in range(n):
        # Simple string length operation (matching Caesar implementation)
        length = len(text)
        count += length
    
    return count

def main():
    if len(sys.argv) != 2:
        print("Usage: python string_ops.py <iterations>")
        sys.exit(1)
    
    try:
        n = int(sys.argv[1])
    except ValueError:
        print("Error: iterations must be an integer")
        sys.exit(1)
    
    if n < 0:
        print("Error: iterations must be non-negative")
        sys.exit(1)
    
    # Choose test based on scale
    if n <= 10000:
        result = string_concatenation(n)
    else:
        result = string_processing(n // 100)
    
    # Don't print result to avoid affecting timing

if __name__ == "__main__":
    main()